src/package-hoister.js

```diff
@@ -242,7 +242,7 @@ export default class PackageHoister {
           continue;
         }
 
-        const isMarkedAsOptional = !depinfo.pkg._reference || this.ignoreOptional;
+        const isMarkedAsOptional = !depinfo.pkg._reference || (depinfo.pkg._reference.optional && this.ignoreOptional);
```

Comment:
i think it'd be more readable to do `depinfo.pkg._reference && depinfo.pkg._reference.optional && this.ignoreOptional`